charmcraft/config.py

```diff
@@ -0,0 +1,176 @@
+    else:
+        # safe fallback
+        msg = error.message
+
+    raise CommandError(msg)
+
+
+@format_checker.checks('url', raises=ValueError)
+def check_url(value):
+    """Check that the URL has at least scheme and net location."""
+    if isinstance(value, str):
+        url = urlparse(value)
+        if url.scheme and url.netloc:
+            return True
+    raise ValueError("must be a full URL (e.g. 'https://some.server.com')")
+
+
+@attr.s(kw_only=True, frozen=True)
+class CharmhubConfig:
+    """Configuration for all Charmhub related options."""
+
+    api_url = attr.ib(default='https://api.staging.charmhub.io')
+    storage_url = attr.ib(default='https://storage.staging.snapcraftcontent.com')
+
+    @classmethod
+    def from_dict(cls, source):
+        """Build from a raw dict."""
+        return cls(**source)
+
+
+class BasicPrime(tuple):
+    """Hold the list of files to include, specified under parts/bundle/prime configs.
+
+    This is a intermediate structure until we have the full Lifecycle in place.
+    """
+
+    @classmethod
+    def from_dict(cls, parts):
+        """Build from a dicts sequence."""
+        prime = parts.get('bundle', {}).get('prime', [])
+
+        # validate that all are relative
+        for spec in prime:
+            # check if it's an absolute path using POSIX's '/' (not os.path.sep, as the charm's
+            # config is independent of the platform where charmcraft is running)
+            if spec[0] == '/':
+                raise CommandError(
+                    "Bad charmcraft.yaml content; the paths specifications in "
+                    "'parts.bundle.prime' must be relative: found {!r}.".format(spec))
+        return cls(prime)
+
+
+@attr.s(kw_only=True, frozen=True)
+class Project:
+    """Configuration for all project-related options, used internally."""
+
+    dirpath = attr.ib(default=None)
+
+
+@attr.s(kw_only=True, frozen=True)
+class Config:
+    """Root of all the configuration."""
+
+    charmhub = attr.ib(default={}, converter=CharmhubConfig.from_dict)
+    parts = attr.ib(default={}, converter=BasicPrime.from_dict)
+    type = attr.ib()
+
+    # this item is provided by the code itself, not the user, as convenience for the
+    # rest of the code
+    project = attr.ib(default=None)
+
+
+CONFIG_SCHEMA = {
+    'type': 'object',
+    'properties': {
+        'type': {'type': 'string', 'enum': ['charm', 'bundle']},
+        'charmhub': {
+            'type': 'object',
+            'properties': {
+                'api_url': {'type': 'string', 'format': 'url'},
+                'storage_url': {'type': 'string', 'format': 'url'},
+            },
+            'additionalProperties': False,
+        },
+        'parts': {
+            'type': 'object',
+            'properties': {
+                'bundle': {
+                    'type': 'object',
+                    'properties': {
+                        'prime': {'type': 'array'},
```

Comment:
shouldn't there be some checks for minItems and types for those items (string)?

Reply:
Added control for these URLs! not only type, but also format (which is guarded now by the schema, and not by the class).

Regarding minItems, I don't see a problem of having an empty list there.

Anyway, this detail can be improved/fixed later if we decide to.